rdata.c

```diff
@@ -845,515 +810,3566 @@ rdata_svcparam_alpn_to_string(buffer_type *output, uint16_t val_len,
+			if(malf1 && !malf2)
+				return 0;
+			return 1;
+		}
+		/* Compare the two fields. */
+		/* If they have a different type field, they are not the
+		 * same. */
+		if(domain1 && !domain2)
+			return 0;
+		if(!domain1 && domain2)
+			return 0;
+		if(domain1 && domain2) {
+			/* Handle RDATA_COMPRESSED_DNAME and
+			 * RDATA_UNCOMPRESSED_DNAME fields. */
+			res = dname_compare(domain_dname(domain1),
+				domain_dname(domain2));
+			if(res != 0)
+				return 0;
+		} else {
+			res = compare_bytestring(rr1->rdata + offset,
+				field_len1, rr2->rdata + offset, field_len2);
+			if(res != 0)
+				return 0;
+		}
+		/* The fields are equal, field_len1 == field_len2. */
+		offset += field_len1;
+	}
+	return 1;
+}
+
+int
+equal_rr_rdata_uncompressed_wire(const nsd_type_descriptor_type *descriptor,
+	const struct rr *rr1, const uint8_t* rr2_rdata, uint16_t rr2_rdlen)
+{
+	size_t i;
+	uint16_t offset1 = 0, offset2 = 0;
+	int res;
+
+	if(!descriptor->has_references) {
+		/* Compare the wireformat of the rdata. */
+		res = compare_bytestring(rr1->rdata, rr1->rdlength,
+			rr2_rdata, rr2_rdlen);
+		if(res != 0)
+			return 0;
+		return 1;
+	}
+
+	for(i=0; i < descriptor->rdata.length; i++) {
+		uint16_t field_len1, field_len2;
+		struct domain* domain1, *domain2;
+		int malf1 = 0, malf2 = 0;
+		/* The fields are equal up to this point. */
+		if((rr1->rdlength == offset1 || rr2_rdlen == offset2) &&
+			descriptor->rdata.fields[i].is_optional) {
+			/* There are no more rdata fields. */
+			/* Check lengths. */
+			int remain1 = rr1->rdlength - offset1;
+			int remain2 = rr2_rdlen - offset2;
+			if(remain1 < remain2)
+				return 0;
+			if(remain1 > remain2)
+				return 0;
+			/* It is equal. */
+			return 1;
+		}
+		if(!lookup_rdata_field_entry(descriptor, i, rr1, offset1,
+			&field_len1, &domain1))
+			malf1 = 1; /* malformed rdata buffer */
+		if(!lookup_rdata_field_entry_uncompressed_wire(descriptor, i,
+			rr2_rdata, rr2_rdlen, offset2, &field_len2, &domain2))
+			malf2 = 1; /* malformed rdata buffer */
+		if(malf1 || malf2) {
+			/* Malformed entries sort last, and are sorted
+			 * equal with other malformed entries. */
+			if(!malf1 && malf2)
+				return 0;
+			if(malf1 && !malf2)
+				return 0;
+			return 1;
+		}
+		/* Compare the two fields. */
+		/* If they have a different type field, they are not the
+		 * same. */
+		if(domain1) {
+			if(domain2) {
+				/* Handle RDATA_COMPRESSED_DNAME and
+				 * RDATA_UNCOMPRESSED_DNAME fields. */
+				res = dname_compare(domain_dname(domain1),
+					domain_dname(domain2));
+				if(res != 0)
+					return 0;
+			} else {
+				if(domain_dname(domain1)->name_size !=
+					buf_dname_length(rr2_rdata+offset2,
+						rr2_rdlen-offset2)) {
+					/* not the same length dnames. */
+					return 0;
+				}
+				if(!dname_equal_nocase(
+					(uint8_t*)dname_name(domain_dname(
+						domain1)),
+					(uint8_t*)rr2_rdata+offset2,
+					rr2_rdlen-offset2)) {
+					/* name comparison not equal. */
+					return 0;
```

Comment:
Should this use the `field_len[12]` or the `buf_dname_length`-result instead of the "remainder" (`rdlen-offset`) for the dname comparisons? Same for the `dname_equal_nocase` in the `else { if(domain2)`  below?

```suggestion
				uint16_t dname_len2 = buf_dname_length(
					rr2_rdata+offset2,
					rr2_rdlen-offset2);
				if(domain_dname(domain1)->name_size !=
					dname_len2) {
					/* not the same length dnames. */
					return 0;
				}
				if(!dname_equal_nocase(
					(uint8_t*)dname_name(domain_dname(
						domain1)),
					(uint8_t*)rr2_rdata+offset2,
					dname_len2)) {
					/* name comparison not equal. */
					return 0;
```

Reply:
I also committed the second case in the if-statement below. That is neater to use the calculated dname length.